napari/plugins/utils.py

```diff
@@ -0,0 +1,160 @@
+from . import PluginError, log_plugin_error
+from . import plugin_manager as napari_plugin_manager
+from ._hookexec import _hookexec
+from typing import Union, List, Set, Optional
+from types import ModuleType
+from ..types import LayerData
+from pluggy.hooks import HookImpl, _HookCaller
+
+HookOrderType = Union[List[str], List[ModuleType], List[HookImpl]]
+
+
+def permute_hook_implementations(
+    hook_caller: _HookCaller, order: HookOrderType
+):
+    """Change the call order of hookimplementations for a pluggy HookCaller.
+
+    Pluggy does not allow a built-in way to change the call order after
+    instantiation.  hookimpls are called in last-in-first-out order.
+    This function accepts the desired call order (a list of plugin names, or
+    plugin modules) and reorders the hookcaller accordingly.
+
+    Parameters
+    ----------
+    hook_caller : pluggy.hooks._HookCaller
+        The hook caller to reorder
+    order : list
+        A list of str, hookimpls, or module_or_class, with the desired
+        CALL ORDER of the hook implementations.
+
+    Raises
+    ------
+    ValueError
+        If the 'order' list cannot be interpreted as a list of "plugin_name"
+        or "plugin" (module_or_class)
+    ValueError
+        if 'order' argument has multiple entries for the same hookimpl
+    """
+    if all(isinstance(o, HookImpl) for o in order):
+        attr = None
+    elif all(isinstance(o, str) for o in order):
+        attr = 'plugin_name'
+    elif any(isinstance(o, str) for o in order):
+        raise TypeError(
+            "order list must be either ALL strings, or ALL modules/classes"
+        )
+    else:
+        attr = 'plugin'
```

Comment:
> @jni said: What is attr?

Reply:
see comment below: this bit is about determining the `type` of the members of the `orders` list.  (due to a perhaps overly permissive function signature).

Reply:
@tlambert03 I still don't quite grasp all the complexity here, but at least some of the comments should be inline here, at the point that the variables are introduced. `attr` is such a generic name...